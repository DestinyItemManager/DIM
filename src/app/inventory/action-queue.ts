import { IPromise } from "angular";

const _queue: Promise<void>[] = [];

// A global queue of functions that will execute one after the other. The function must return a promise.
// fn is either a blocking function or a function that returns a promise
export function queueAction(fn: () => Promise<void> | IPromise<void>) {
  let promise = (_queue.length) ? _queue[_queue.length - 1] : Promise.resolve();
  // Execute fn regardless of the result of the existing promise. We
  // don't use finally here because finally can't modify the return value.
  promise = promise.then(() => {
    return fn();
  }, () => {
    return fn();
  }).then((value) => {
    _queue.shift();
    return value;
  }, (e) => {
    _queue.shift();
    throw e;
  });
  _queue.push(promise);

  return promise;
}

// Wrap a function to produce a function that will be queued when invoked
export function queuedAction(fn, context?) {
  return (...args) => {
    return queueAction(() => {
      return fn.apply(context, args);
    });
  };
}
