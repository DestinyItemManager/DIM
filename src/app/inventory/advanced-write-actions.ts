import { DestinyAccount } from "../accounts/destiny-account.service";
import { DimItem, DimSocket } from "./store/d2-item-factory.service";
import { AwaType, AwaAuthorizationResult, AwaUserSelection, insertSocketPlug, DestinySocketArrayType } from "bungie-api-ts/destiny2";
import { requestAdvancedWriteActionToken } from "../bungie-api/destiny2-api";
import * as idbKeyval from 'idb-keyval';
import { httpAdapterWithRetry } from "../bungie-api/bungie-service-helper";

let awaCache: {
  [key: number]: AwaAuthorizationResult & { used: number };
};

// TODO: we need an interface for presenting non-reusable plugs like mods and shaders
// TODO: owner can't be "vault" I bet
export async function insertPlug(account: DestinyAccount, item: DimItem, socket: DimSocket, plugItemHash: number) {
  const actionToken = await getAwaToken(account, AwaType.InsertPlugs, item);

  return insertSocketPlug(httpAdapterWithRetry, {
    actionToken,
    itemInstanceId: item.id,
    plug: {
      socketIndex: socket.socketIndex,
      socketArrayType: DestinySocketArrayType.Default,
      plugItemHash
    },
    characterId: item.owner,
    membershipType: account.platformType
  });
}

/**
 * Given a request for an action token, and a particular action type, return either
 * a cached token or fetch and return a new one.
 *
 * @param item The item is optional unless the type is DismantleGroupA, but it's best to pass it when possible.
 */
export async function getAwaToken(account: DestinyAccount, action: AwaType, item?: DimItem): Promise<string> {
  if (!awaCache) {
    // load from cache first time
    awaCache = (await idbKeyval.get('awa-tokens') || {}) as {
      [key: number]: AwaAuthorizationResult & { used: number };
    };
  }

  let info = awaCache[action];
  if (!info || !tokenValid(info)) {
    try {
      info = awaCache[action] = {
        ...await requestAdvancedWriteActionToken(account, action, item),
        used: 0
      };

      // Deletes of "group A" require an item and shouldn't be cached
      if (action === AwaType.DismantleGroupA) {
        delete awaCache[action]; // don't cache
      }

      // TODO: really should use a separate db for this
      // without blocking, save this
      idbKeyval.set('awa-tokens', awaCache);
    } catch (e) {
      throw new Error("Unable to get a token: " + e.message);
    }

    if (!info || !tokenValid(info)) {
      throw new Error("Unable to get a token: " + info ? info.developerNote : "no response");
    }
  }

  info.used++;
  return info.actionToken;
}

function tokenValid(info: AwaAuthorizationResult & { used: number }) {
  return (!info.validUntil || new Date(info.validUntil) > new Date()) &&
    (info.maximumNumberOfUses === 0 && info.used <= info.maximumNumberOfUses) &&
    info.userSelection === AwaUserSelection.Approved;
}
