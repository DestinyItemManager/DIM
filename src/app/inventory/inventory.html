<div class="sticky-header-background" scroll-class="something-is-sticky"></div>
<dim-stores class="stores" stores="$ctrl.stores"></dim-stores>
<loadout-drawer stores="$ctrl.stores" account="$ctrl.account"></loadout-drawer>
<dim-compare></dim-compare>
<dim-item-discuss></dim-item-discuss>
<div id="drag-help" class="drag-help-hidden" ng-i18next="Help.Drag"></div>
<dim-farming></dim-farming>
<dim-clear-new-items></dim-clear-new-items>
<random-loadout></random-loadout>
