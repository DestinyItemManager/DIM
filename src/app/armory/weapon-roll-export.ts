import { DimItem } from 'app/inventory/item-types';
import { getMasterworkStatNames } from 'app/utils/item-utils';
import { getWeaponArchetypeSocket, isWeaponMasterworkSocket } from 'app/utils/socket-utils';
import { ItemCategoryHashes } from 'data/d2/generated-enums';

/**
 * Exports a weapon roll as human-readable text format suitable for sharing
 * on platforms like /r/sharditkeepit
 */
export function formatWeaponRollForExport(item: DimItem): string {
  if (!item.bucket.inWeapons || !item.sockets) {
    return '';
  }

  const lines: string[] = [];

  // Weapon name
  lines.push(item.name);
  lines.push('');

  // Intrinsic Perk / Frame (Archetype)
  const archetypeSocket = getWeaponArchetypeSocket(item);
  if (archetypeSocket?.plugged?.plugDef.displayProperties.name) {
    lines.push(`* ${archetypeSocket.plugged.plugDef.displayProperties.name}`);
  }

  // Iterate through all perk sockets in order
  if (item.sockets) {
    for (const socket of item.sockets.allSockets) {
      // Only include perk sockets, excluding weapon mods, masterwork, and archetype
      if (
        socket.isPerk &&
        socket.plugOptions.length > 0 &&
        !isWeaponMasterworkSocket(socket) &&
        socket !== archetypeSocket
      ) {
        // Get all plug options for this socket, excluding weapon mod damage
        const plugNames = socket.plugOptions
          .filter(
            (plug) =>
              plug.plugDef.displayProperties.name &&
              !plug.plugDef.itemCategoryHashes?.includes(ItemCategoryHashes.WeaponModsDamage),
          )
          .map((plug) => plug.plugDef.displayProperties.name)
          .filter(Boolean);

        if (plugNames.length > 0) {
          lines.push(`* ${plugNames.join(' / ')}`);
        }
      }
    }
  }

  // Masterwork (if applicable)
  const masterworkStatName = getMasterworkStatNames(item.masterworkInfo);
  if (masterworkStatName) {
    lines.push('');
    lines.push(`* ${masterworkStatName}`);
  }

  return lines.join('\n');
}

/**
 * Copy text to clipboard using the modern Clipboard API
 */
export async function copyToClipboard(text: string): Promise<boolean> {
  try {
    if (navigator.clipboard && window.isSecureContext) {
      await navigator.clipboard.writeText(text);
      return true;
    }
    return false;
  } catch (error) {
    // eslint-disable-next-line no-console
    console.error('Failed to copy to clipboard:', error);
    return false;
  }
}

/**
 * Check if clipboard functionality is available
 */
export function isClipboardAvailable(): boolean {
  return Boolean(navigator.clipboard && window.isSecureContext);
}
