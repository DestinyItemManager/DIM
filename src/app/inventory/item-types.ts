import {
  DestinyStat,
  DestinyStatDefinition,
  DestinyInventoryItemDefinition,
  DestinySandboxPerkDefinition,
  DestinyObjectiveProgress,
  DestinySocketCategoryDefinition,
  DestinyClass,
  DestinyItemTierTypeInfusionBlock,
  DestinyItemQualityBlockDefinition
} from 'bungie-api-ts/destiny2';
import { DimItemInfo } from './dim-item-info';
import { DimStore, StoreServiceType, D1StoreServiceType, D2StoreServiceType } from './store-types';
import { InventoryBucket } from './inventory-buckets';
import { D2RatingData } from '../item-review/d2-dtr-api-types';
import { D1RatingData } from '../item-review/d1-dtr-api-types';
import { DtrRating } from '../item-review/dtr-api-types';

/**
 * A generic DIM item, representing almost anything. Use this type when you can handle both D1 and D2 items,
 * or you don't use anything specific to one of them.
 */
// TODO: This interface is clearly too large - break out interfaces for common subsets
export interface DimItem {
  /** The ID of the store that currently contains this item. */
  owner: string;
  /** The version of Destiny this comes from. */
  destinyVersion: 1 | 2;
  /** The bucket the item is currently in. */
  location: InventoryBucket;
  /** The bucket the item normally resides in (even though it may currently be elsewhere, such as in the postmaster). */
  bucket: InventoryBucket;
  /** The inventoryItemHash, see DestinyInventoryItemDefinition. */
  hash: number;
  /** This is the type of the item (see InventoryBuckets) regardless of location. This string is a DIM concept with no direct correlation to the API types. */
  type: string;
  /** String names of "item categories" the item may be in. These are not the same as DestinyItemCategoryDefinitions. */
  categories: string[];
  /** A readable English name for the rarity of the item (e.g. "Exotic", "Rare"). */
  tier: string;
  /** Is this an Exotic item? */
  isExotic: boolean;
  /** Did this come from a vendor instead of character inventory? */
  isVendorItem: boolean;
  /** Localized name of the item. */
  name: string;
  /** Localized description of the item. */
  description: string;
  /** Icon path for the item. */
  icon: string;
  /** Some items have a secondary icon, namely Emblems. */
  secondaryIcon: string;
  /** Whether this item CANNOT be transferred. */
  notransfer: boolean;
  /** Whether we can pull this item from the postmaster */
  canPullFromPostmaster: boolean;
  /** Item instance id. Non-instanced items have id "0" for D1 compatibility. */
  id: string;
  /** Is this item currently equipped? */
  equipped: boolean;
  /** Is this "equipment" (items that can be equipped). */
  equipment: boolean;
  /**
   * If defined, this is the label used to check if the character has other items of
   * matching types already equipped.
   *
   * For instance, when you aren't allowed to equip more than one Exotic Weapon, that's
   * because all exotic weapons have identical labels and the game checks the
   * to-be-equipped item's label vs. all other already equipped items (other
   * than the item in the slot that's about to be occupied).
   */
  equippingLabel?: string;
  /** Is this item complete (leveled, unlocked, objectives complete). */
  complete: boolean;
  /** How many items does this represent? Only greater than one if maxStackSize is greater than one. */
  amount: number;
  /** The primary stat (Attack, Defense, Speed) of the item. */
  primStat: DestinyStat | null;
  /** Localized name of this item's type. */
  typeName: string;
  /** The level a character must be to equip this item. */
  equipRequiredLevel: number;
  /** The maximum number of items that stack together for this item type. */
  maxStackSize: number;
  /** The class this item is restricted to. Unknown means it can be used by any class. */
  classType: DestinyClass;
  /** The readable English name of the class this item is restricted to. */
  classTypeName: string;
  /** The localized name of the class this item is restricted to. */
  classTypeNameLocalized: string;
  /** The readable name of the damage type associated with this item. */
  dmg: 'kinetic' | 'arc' | 'solar' | 'void';
  /** Whether this item should be shown. */
  visible: boolean;
  /** Whether this item can be locked. */
  lockable: boolean;
  /** Is this item tracked? (D1 quests/bounties). */
  tracked: boolean;
  /** Is this item locked? */
  locked: boolean;
  /** Is this a masterwork? (D2 only) */
  masterwork: boolean;
  /** Is this classified? Some items are classified in the manifest. */
  classified: boolean;
  /** What percent complete is this item (considers XP and objectives). */
  percentComplete: number;
  /** Should we hide the percentage display? */
  hidePercentage: boolean;
  /** Can this be tagged? */
  taggable: boolean;
  /** Can this be compared with other items? */
  comparable: boolean;
  /** Can this be reviewed? */
  reviewable: boolean;
  /** Is this a new item? */
  isNew: boolean;
  /** DIM tagging and notes info. */
  dimInfo: DimItemInfo;
  /** The "base power" without any power-enhancing mods. */
  basePower: number;
  /** A synthetic unique ID used to help Angular tell items apart. This changes to signal that Angular should re-render the item. */
  index: string;
  /** Can this be infused? */
  infusable: boolean;
  /** Can this be used as infusion fuel? */
  infusionFuel: boolean;
  /** The talent grid, used for D1 perks and D1/D2 subclass grids. */
  talentGrid: DimTalentGrid | null;
  /** Perks, which are specifically called-out special abilities of items shown in the game's popup UI. */
  perks: DimPerk[] | null;
  /** Detailed stats for the item. */
  stats: DimStat[] | null;
  /** Any objectives associated with the item. */
  objectives: DimObjective[] | null;
  /** Is this an engram? */
  isEngram: boolean;
  /** The reference hash for lore attached to this item (D2 only). */
  loreHash: number;
  /** A timestamp of when, in this session, the item was last manually moved */
  lastManuallyMoved: number;

  /** Information about community ratings. */
  dtrRating: DtrRating | null;

  /** Can this item be equipped by the given store? */
  canBeEquippedBy(store: DimStore): boolean;
  /** Is this item in the given category name? Only really useful in D1. */
  inCategory(categoryName: string): boolean;
  /** Could this be added to a loadout? */
  canBeInLoadout(): boolean;
  /** "Touch" the item to mark it as having been manually moved. */
  updateManualMoveTimestamp(): void;
  /** The Stores service associated with this item. */
  getStoresService(): StoreServiceType;

  /** Check if this item is from D1. Inside an if statement, this item will be narrowed to type D1Item. */
  isDestiny1(): this is D1Item;
  /** Check if this item is from D2. Inside an if statement, this item will be narrowed to type D2Item.s */
  isDestiny2(): this is D2Item;
}

/**
 * A Destiny 1 item. Use this type when you need specific D1 properties.
 */
export interface D1Item extends DimItem {
  primStat: D1PrimStat | null;
  talentGrid: D1TalentGrid | null;
  /** The overall item group (e.g. Weapons, Armor) this item is in. See InventoryBuckets. */
  sort?: string;
  stats: D1Stat[] | null;
  /** Armor quality evaluation. */
  quality: {
    /** The maximum stat range this armor could achieve when fully infused. */
    min: number;
    /** The minimum stat range this armor could achieve when fully infused. */
    max: number;
  } | null;
  /** Which D1 "year" this item was associated with. */
  year: number;
  /** Hashes that allow us to figure out where this item can be found (what activities, locations, etc.) */
  sourceHashes: number[];
  /** Can this item be tracked? (For quests/bounties.) */
  trackable: boolean;

  dtrRating: D1RatingData | null;
  getStoresService(): D1StoreServiceType;
}

/**
 * A Destiny 2 item. Use this type when you need specific D2 properties.
 */
export interface D2Item extends DimItem {
  primStat: D2PrimStat | null;
  /** D2 items use sockets and plugs to represent everything from perks to mods to ornaments and shaders. */
  sockets: DimSockets | null;
  /** Some items have a "flavor objective", such as emblems that track stats. */
  flavorObjective: DimFlavorObjective | null;
  /** If this item is a masterwork, this will include information about its masterwork properties. */
  masterworkInfo: DimMasterwork | null;
  /** Information about how this item works with infusion. */
  infusionQuality: DestinyItemQualityBlockDefinition | null;
  /** More infusion information about what can be infused with the item. */
  infusionProcess: DestinyItemTierTypeInfusionBlock | null;
  /** The DestinyVendorDefinition hash of the vendor that can preview the contents of this item, if there is one. */
  previewVendor?: number;
  dtrRating: D2RatingData | null;
  getStoresService(): D2StoreServiceType;
}

export interface D2PrimStat extends DestinyStat {
  stat: DestinyStatDefinition & { statName: string };
}
export interface D1PrimStat extends DestinyStat {
  stat: any;
}

export interface DimMasterwork {
  /** How much has the masterwork objective been progressed? This is usually kill counting. */
  progress?: number;
  /** Which type of masterwork is it? */
  typeName: 'Vanguard' | 'Crucible' | null;
  /** The icon associated with the type. */
  typeIcon: string;
  /** The localized description associated with the type. */
  typeDesc: string | null;
  /** The stat that is enhanced by this masterwork. */
  statHash?: number;
  /** The name of the stat enhanced by this masterwork. */
  statName: string;
  /** How much the stat is enhanced by this masterwork. */
  statValue?: number;
}

export interface DimStat {
  /** Base stat without "increase defense", etc. Bonuses applied. */
  base: number;
  /** Stat bonus from leveling this item. A D1 thing. */
  bonus: number;
  /** DestinyStatDefinition hash. */
  statHash: number;
  /** Localized stat name. */
  name: string;
  /** Stat identifier. D1 only. */
  id: number;
  /** Sort order. */
  sort: number;
  /** Absolute stat value. */
  value?: number;
  /** The maximum value this stat can have. */
  maximumValue: number;
  /** Should this be displayed as a bar or just a number? */
  bar: boolean;
  /** Is this a placeholder for a "missing" stat (for compare view) */
  missingStat?: boolean;
}

export interface D1Stat extends DimStat {
  scaled?: {
    max: number;
    min: number;
  };
  split?: number;
  qualityPercentage?: {
    max: number;
    min: number;
    range: string;
  };
}

export interface DimObjective {
  /** Localized display of the objective status. */
  displayName: string;
  /** Localized description of the objective. */
  description?: string;
  /** Current value. */
  progress: number;
  /** Value at which this objective would be considered "complete". */
  completionValue: number;
  /** Is this complete? */
  complete: boolean;
  /** Is this a checkbox? */
  boolean: boolean;
  /** The actual string to display for the objective value (e.g "5/10" or "50%") */
  display?: string;
  /** Override display styles for objectives, such as 'trials' or 'integer' */
  // TODO: fold 'boolean' into this
  displayStyle: string | null;
}

export interface DimFlavorObjective {
  /** Localized description for the flavor objective. */
  description: string;
  /** Icon for the flavor objective. */
  icon: string;
  /** Current value of the tracker. */
  progress: number;
}

export interface DimTalentGrid {
  /** A flat list of nodes in the grid. */
  nodes: DimGridNode[];
  /** Is the grid complete (leveled and unlocked)? */
  complete: boolean;
}

export interface D1TalentGrid extends DimTalentGrid {
  nodes: D1GridNode[];
  /** Have all nodes been unlocked via XP? */
  xpComplete: boolean;
  /** How much XP in total to unlock the grid fully? */
  totalXPRequired: number;
  /** Total XP earned in the grid. */
  totalXP: number;
  /** Is there an "Ascend" action node? */
  hasAscendNode: boolean;
  /** Has this item been ascended? */
  ascended: boolean;
  /** Can this item be infused? */
  infusable: boolean;
  /** A representation of the perk grid for DTR integration. */
  dtrPerks: string;
  /** Another representation of the perk grid for DTR integration. */
  dtrRoll: string;
}

export interface DimGridNode {
  /** Localized name of the grid. */
  name: string;
  /** Talent grid definition hash. */
  hash: number;
  /** Localized description of the grid. */
  description: string;
  /** Icon of the grid. */
  icon: string;
  /** Column position in the grid. */
  column: number;
  /** Row position in the grid. */
  row: number;
  /** Is the node selected (lit up in the grid) */
  activated: boolean;
  /** The item level at which this node can be unlocked */
  activatedAtGridLevel: number;
  /** Only one node in this column can be selected (scopes, etc) */
  exclusiveInColumn: boolean;
  /** Whether or not the material cost has been paid for the node */
  unlocked: boolean;
  /** Some nodes don't show up in the grid, like purchased ascend nodes */
  hidden: boolean;
}

export interface D1GridNode extends DimGridNode {
  /** Is this an ornament node? */
  ornament: boolean;
  /** How much XP has been put into this node? */
  xp: number;
  /** How much XP is required to unlock the node? */
  xpRequired: number;
  /** Has the XP requirement been met? */
  xpRequirementMet: boolean;

  /** A representation of the node for DTR integration. */
  dtrHash: string | null;
  /** Another representation of the node for DTR integration. */
  dtrRoll: string;

  /** Is this the best perk as chosen by DTR ratings? */
  bestRated?: boolean;
}

/**
 * Dim's view of a "Plug" - an item that can go into a socket.
 * In D2, both perk grids and mods/shaders are sockets with plugs.
 */
export interface DimPlug {
  /** The item associated with this plug. */
  plugItem: DestinyInventoryItemDefinition;
  /** Is this the best perk as chosen by DTR ratings? */
  bestRated?: boolean;
  /** Perks associated with the use of this plug. */
  perks: DestinySandboxPerkDefinition[];
  /** Objectives associated with this plug, usually used to unlock it. */
  plugObjectives: DestinyObjectiveProgress[];
  /** Is the plug enabled? For example, some perks only activate on certain planets. */
  enabled: boolean;
  /** If not enabled, this is the localized reasons why, as a single string. */
  enableFailReasons: string;
  /** Is this a Masterwork plug? */
  isMasterwork: boolean;
}

export interface DimSocket {
  /** The index of this socket in the overall socket list, used for the AWA InsertPlug API. */
  socketIndex: number;
  /** The currently inserted plug item, if any. */
  plug: DimPlug | null;
  /** Potential plugs for this socket. */
  plugOptions: DimPlug[];
}

export interface DimSocketCategory {
  /** A grouping of sockets. */
  category: DestinySocketCategoryDefinition;
  /** The sockets in the group. */
  sockets: DimSocket[];
}

export interface DimSockets {
  /** A flat list of all sockes on the item. */
  sockets: DimSocket[];
  /** Sockets grouped by category. */
  categories: DimSocketCategory[];
}

export interface DimPerk extends DestinySandboxPerkDefinition {
  /** Localized reason for why the perk can't be used. */
  requirement: string;
}
