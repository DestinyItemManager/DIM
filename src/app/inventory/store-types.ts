import {
  DestinyClass,
  DestinyProgression,
  DestinyCharacterComponent,
  DestinyFactionDefinition
} from 'bungie-api-ts/destiny2';
import { Loadout } from '../loadout/loadout.service';
import { D1ManifestDefinitions } from '../destiny1/d1-definitions.service';
import { D2ManifestDefinitions } from '../destiny2/d2-definitions.service';
import { IPromise } from 'angular';
import { DimItem, D2Item, D1Item } from './item-types';
import { DestinyAccount } from '../accounts/destiny-account.service';
import { ConnectableObservable } from 'rxjs/observable/ConnectableObservable';
import { InventoryBucket } from './inventory-buckets';

/**
 * A generic store service that produces stores and items that are the same across D1 and D2. Use this
 * if you don't care about the differences between the two.
 */
export interface StoreServiceType<StoreType = DimStore, VaultType = DimVault, ItemType = DimItem> {
  /** Get the active or last-played character. */
  getActiveStore(): StoreType | undefined;
  /** Get a list of all characters plus the vault. */
  getStores(): StoreType[];
  /** Get a store by character ID. */
  getStore(id: string): StoreType | undefined;
  /** Get the vault. */
  getVault(): VaultType | undefined;
  /** Get all items across all stores. */
  getAllItems(): ItemType[];
  /** A stream of store updates for a particular account. */
  getStoresStream(account: DestinyAccount): ConnectableObservable<StoreType[] | undefined>;
  /** Get an item matching certain characteristics, no matter where it is in inventory. */
  getItemAcrossStores(params: {
    id?: string;
    hash?: number;
    notransfer?: boolean;
    amount?: number;
  }): ItemType | undefined;
  /** Refresh just character info (current light/stats, etc.) */
  updateCharacters(account?: DestinyAccount): IPromise<StoreType[]>;
  /** Reload inventory completely. */
  reloadStores(): Promise<StoreType[] | undefined>;
  /** Reload DTR rating data. */
  refreshRatingsData(): void;

  /** Tell Redux things have changed. Temporary bridge for Redux. */
  touch(): void;
}

/**
 * A Destiny 2 store service. This will use D2 types everywhere, avoiding the need to check.
 */
export type D2StoreServiceType = StoreServiceType<D2Store, D2Vault, D2Item>;

/**
 * A Destiny 1 store service. This will use D1 types everywhere, avoiding the need to check.
 */
export type D1StoreServiceType = StoreServiceType<D1Store, D1Vault, D1Item>;

/**
 * A generic DIM character or vault - a "store" of items. Use this type when you can handle both D1 and D2 characters,
 * or you don't use anything specific to one of them.
 */
export interface DimStore {
  /** An ID for the store. Character ID or 'vault'. */
  id: string;
  /** Localized name for the store. */
  name: string;
  /** All items in the store, across all buckets. */
  items: DimItem[];
  /** All items, grouped by their bucket. */
  buckets: { [bucketId: string]: DimItem[] };
  /** The Destiny version this store came from. */
  destinyVersion: 1 | 2;
  /** An icon (emblem) for the store. */
  icon: string;
  /** Is this the most-recently-played character? */
  current: boolean;
  /** The date the character was last played. */
  lastPlayed: Date;
  /** Emblem background. */
  background: string;
  /** Character level. */
  level: number;
  /** Progress towards the next level (or "prestige level") */
  percentToNextLevel: number;
  /** Power/light level. */
  powerLevel: number;
  /** String class name. */
  class: 'titan' | 'warlock' | 'hunter';
  /** Integer class type. */
  classType: DestinyClass;
  /** Localized class name. */
  className: string;
  /** Localized gender. */
  gender: string;
  /** Localized gender and race together. */
  genderRace: string;
  /** Is this the vault? */
  isVault: boolean;
  /** Character stats. */
  stats: {};
  /** Character progression. */
  progression: null | {
    progressions: DestinyProgression[];
  };

  /** Apply updated character info. */
  updateCharacterInfo(
    defs: D1ManifestDefinitions | D2ManifestDefinitions,
    bStore: any
  ): IPromise<DimStore[]>;

  /**
   * Get the total amount of this item in the store, across all stacks,
   * excluding stuff in the postmaster.
   */
  amountOfItem(item: DimItem): number;
  /**
   * How much of items like this item can fit in this store? For
   * stackables, this is in stacks, not individual pieces.
   */
  capacityForItem(item: DimItem): number;
  /**
   * How many *more* items like this item can fit in this store?
   * This takes into account stackables, so the answer will be in
   * terms of individual pieces.
   */
  spaceLeftForItem(item: DimItem): number;

  /** Remove an item from this store. Returns whether it actually removed anything. */
  removeItem(item: DimItem): boolean;

  /** Add an item to the store. */
  addItem(item: DimItem): void;

  /** Create a loadout from this store's equipped items. */
  loadoutFromCurrentlyEquipped(name: string): Loadout;

  /** Check if this store is from D1. Inside an if statement, this item will be narrowed to type D1Store. */
  isDestiny1(): this is D1Store;
  /* Check if this store is from D2. Inside an if statement, this item will be narrowed to type D2Store. */
  isDestiny2(): this is D2Store;

  /** The stores service associated with this store. */
  getStoresService(): StoreServiceType;

  /** A temporary way of telling Redux that something about the stores has changed. */
  touch(): void;
}

/** How many items are in each vault bucket. DIM hides the vault bucket concept from users but needs the count to track progress. */
interface VaultCounts {
  [bucketId: number]: { count: number; bucket: InventoryBucket };
}

export interface DimVault extends DimStore {
  vaultCounts: VaultCounts;
  legendaryMarks: number;
  glimmer: number;
  silver: number;
}

export interface D1Vault extends D1Store {
  vaultCounts: VaultCounts;
  legendaryMarks: number;
  glimmer: number;
  silver: number;
}

export interface D2Vault extends D2Store {
  vaultCounts: VaultCounts;
  legendaryMarks: number;
  glimmer: number;
  silver: number;
}

export interface D2CharacterStat {
  /** The DestinyStatDefinition hash for the stat. */
  id: number;
  /** The localized name of the stat. */
  name: string;
  /** The localized description of the stat. */
  description: string;
  /** The current value of the stat. */
  value: number | string;
  /** An icon associated with the stat. */
  icon: string;
  /** The size of one "tier" of the stat (for D1 stats) */
  tierMax?: number;
  /** The stat divided into tiers. Each element is how full that tier is. */
  tiers?: number[] | string[];
  /** Whether this stat is inaccurate because it relies on classified items (like base power). */
  hasClassified?: boolean;
}

export interface D1CharacterStat {
  /** Stat identifier (e.g. "STAT_INTELLECT") */
  id: string;
  /** The localized name of the stat. */
  name?: string;
  /** An icon associated with the stat. */
  icon?: string;
  /** The current value of the stat. */
  value: number;

  /** A localized description of this stat's effect. */
  effect?: string;
  /** The stat value, clamped to <300. TODO: what is this? */
  normalized?: number;
  /** Which tier (out of 5) has been activated. */
  tier?: number;
  /** The size of one "tier" of the stat (for D1 stats) */
  tierMax?: number;
  /** The stat divided into tiers. Each element is how full that tier is. */
  tiers?: number[];
  /** TODO: remove this and normalized */
  remaining?: number;
  /** Cooldown time for the associated ability. */
  cooldown?: string;
  /** Percentage of maximum stat value. */
  percentage?: string;
}

export interface D1Progression extends DestinyProgression {
  /** The faction definition associated with this progress. */
  faction: DestinyFactionDefinition & {
    factionName: string;
    factionIcon: string;
  };
  order: number;
}

/**
 * A D1 character. Use this when you need D1-specific properties or D1-specific items.
 */
export interface D1Store extends DimStore {
  items: D1Item[];
  buckets: { [bucketId: string]: D1Item[] };
  stats: {
    [hash: string]: D1CharacterStat;
  };
  progression: null | {
    progressions: D1Progression[];
  };

  // TODO: shape?
  advisors: any;

  updateCharacterInfo(
    defs: D1ManifestDefinitions,
    bStore: any
  ): IPromise<D1Store[]>;
  updateCharacterInfoFromEquip(characterInfo);
  /** Which faction is this character currently aligned with? */
  factionAlignment();
  getStoresService(): D1StoreServiceType;
}

/**
 * A D2 character. Use this when you need D2-specific properties or D2-specific items.
 */
export interface D2Store extends DimStore {
  items: D2Item[];
  buckets: { [bucketId: string]: D2Item[] };
  /** The vault associated with this store. */
  vault?: D2Vault;
  stats: {
    maxBasePower?: D2CharacterStat;
    [statHash: number]: D2CharacterStat;
  };
  updateCharacterInfo(
    defs: D2ManifestDefinitions,
    bStore: DestinyCharacterComponent
  ): IPromise<D2Store[]>;
  getStoresService(): D1StoreServiceType;
}
