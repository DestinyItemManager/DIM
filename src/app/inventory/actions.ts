import { createStandardAction } from 'typesafe-actions';
import { DimStore } from './store-types';
import { DimItem } from './item-types';
import { InventoryBuckets } from './inventory-buckets';
import { DimItemInfo } from './dim-item-info';

/**
 * Reflect the old stores service data into the Redux store as a migration aid.
 */
export const update = createStandardAction('inventory/UPDATE')<{
  stores: DimStore[];
  buckets?: InventoryBuckets;
  newItems?: Set<string>;
}>();

/**
 * Set the bucket info.
 */
export const setBuckets = createStandardAction('inventory/SET_BUCKETS')<InventoryBuckets>();

/**
 * Move an item from one store to another.
 */
export const moveItem = createStandardAction('inventory/MOVE_ITEM')<{
  item: DimItem;
  source: DimStore;
  target: DimStore;
  equip: boolean;
  amount: number;
}>();

// TODO: tags/notes should probably be their own part of state
export const setTag = createStandardAction('inventory/SET_TAG')<{
  itemId: string;
  tag: string;
}>();

/** Update the set of new items. */
export const setNewItems = createStandardAction('new_items/SET')<Set<string>>();

/** Update the item infos (tags/notes). */
export const setTagsAndNotes = createStandardAction('tag_notes/SET')<{
  [key: string]: DimItemInfo;
}>();

/** Set the tags/notes for a single item. */
export const setTagsAndNotesForItem = createStandardAction('tag_notes/UPDATE_ITEM')<{
  key: string;
  info: DimItemInfo;
}>();

/** Notify that a stackable stack has begun or ended dragging. A bit overkill to put this in redux but eh. */
export const stackableDrag = createStandardAction('stackable_drag/DRAG')<boolean>();
/** Notify that a stackable stack has started or stopped hovering. */
export const stackableHover = createStandardAction('stackable_drag/HOVER')<boolean>();
