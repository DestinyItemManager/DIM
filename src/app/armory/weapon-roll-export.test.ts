import { DimItem } from 'app/inventory/item-types';
import { DimStore } from 'app/inventory/store-types';
import { getTestStores } from 'testing/test-utils';
import { formatWeaponRollForExport } from './weapon-roll-export';

describe('formatWeaponRollForExport', () => {
  let stores: DimStore[];

  beforeAll(async () => {
    stores = await getTestStores();
    expect(stores).toBeDefined();
  });

  /** Get a test item from the stores based on a predicate */
  function getTestItem(itemPredicate: (item: DimItem) => boolean): DimItem {
    const item = stores.flatMap((s) => s.items).find(itemPredicate);
    expect(item).toBeDefined();
    return item!;
  }

  it('should return empty string for non-weapon items', () => {
    const armor = getTestItem((i) => Boolean(i.bucket.inArmor));
    const result = formatWeaponRollForExport(armor);
    expect(result).toBe('');
  });

  it('should format a weapon roll correctly', () => {
    const weapon = getTestItem((i) => Boolean(i.bucket.inWeapons && i.sockets));
    expect(weapon).toBeDefined();

    const result = formatWeaponRollForExport(weapon);
    expect(result).toMatchSnapshot();
  });

  it('should include masterwork stat when available', () => {
    const weapon = getTestItem((i) =>
      Boolean(i.bucket.inWeapons && i.masterworkInfo?.stats?.length),
    );
    expect(weapon).toBeDefined();
    expect(weapon.masterworkInfo?.stats?.length).toBeGreaterThan(0);

    const result = formatWeaponRollForExport(weapon);
    expect(result).toMatchSnapshot();
  });
});
