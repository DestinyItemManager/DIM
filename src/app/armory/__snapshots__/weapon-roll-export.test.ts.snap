// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`formatWeaponRollForExport should format a weapon roll correctly 1`] = `
"Quicksilver Storm

* Rocket Tracers
* Corkscrew Rifling
* High-Caliber Rounds
* Grenade Chaser
* Hand-Laid Stock
* Kill Tracker / Crucible Tracker"
`;

exports[`formatWeaponRollForExport should include masterwork stat when available 1`] = `
"Pizzicato-22

* Adaptive Frame
* Corkscrew Rifling / Full Bore
* Alloy Magazine / Ricochet Rounds
* Ambitious Assassin
* Multikill Clip
* Suros Synergy
* Kill Tracker / Crucible Tracker

* Reload Speed"
`;
